### ISSUE TITLE
Change API to make configureDragDrop static

### ISSUE THREAD

**Issue by gaearon** [gaearon/react-dnd#54]
I propose we make `configureDragDrop` static, and pass current `component` as the first argument to each method. This change is a stepping stone to fix `endDrag` not firing on unmounted component (#38, we'll pass `null` as `component`) and also a stepping stone to allow “resurfacing” drag sources (#53, we'll pass another component as `component`).

I'd like this change to be implemented in a backward-compatible way with a deprecation warning. We will make it a breaking change in 1.0 with #48.

- [ ] Support static `configureDragDrop` and pass `component` as the first argument to all methods
- [ ] Update the docs to use the new way
- [ ] Add a deprecation warning with upgrade instructions but still allow old-style `configureDragDrop` definitions (without `component` parameter)
- [ ] Make sure deprecation warning is logged once per class and includes class name (like React's own warnings)
- [ ] Prepare a separate PR that removes deprecation warning and makes it a breaking change. We will do that on `1.0` branch.

Any comments or volunteers?

@hakanderyal @nelix

**Comment by gaearon** [comment]
To clarify. Before:

```js
var Card = React.createClass({
  mixins: [DragDropMixin],

  configureDragDrop(registerType) {
    registerType(ItemTypes.CARD, {
      dragSource: {
        beginDrag() {
          return {
            item: {
              id: this.props.id,
              children: <Card {...this.props}/>
            }
          };
        }
      },

      dropTarget: {
        over(item) {
          this.props.moveCard(item.id, this.props.id);
        }
      }
    });
  },
```

After:

```js
var Card = React.createClass({
  mixins: [DragDropMixin],

  statics: {
    configureDragDrop(registerType) {
      registerType(ItemTypes.CARD, {
        dragSource: {
          beginDrag(component) {
            return {
              item: {
                id: component.props.id,
                children: <Card {...component.props}/>
              }
            };
          }
        },

        dropTarget: {
          over(component, item) {
            component.props.moveCard(item.id, component.props.id);
          }
        }
      });
    }
  }
```

We may want to have a less nested API later, but this is not the point. The point is using `component` instead of `this` so we can later be free to pass `null` or another component if we wish so.

This breaks the use case when `this` is accessed directly in `configureDragDrop` but it's a rare use case and seems only useful in [contrived examples](https://github.com/gaearon/react-dnd/blob/master/examples/_dustbin-interesting/Item.js#L23).

**Comment by gaearon** [comment]
Released in `react-dnd@0.6.4`.
